Add group-level reports for an exercise/resource learners

MisRob opened the pull request:
### Summary

Add "View by groups" feature to exercise learners list reports

#### Screenshots

<img width="734" alt="Screenshot 2019-05-05 at 22 04 32" src="https://user-images.githubusercontent.com/13509191/57199661-e1b96a00-6f81-11e9-807b-b3699c2099fa.png">
<img width="734" alt="Screenshot 2019-05-05 at 22 04 40" src="https://user-images.githubusercontent.com/13509191/57199682-0d3c5480-6f82-11e9-8519-7df4d4900291.png">


#### Design files

- #5403 screens
- https://www.figma.com/file/D5wXhmWNGyt60SpGBgFucwZU/Coach-reports-0.12?node-id=1385%3A4822I'

#### Notes

- We'll need to discuss if/how to visually treat edge cases like empty groups (only non-empty ones are displayed now) and learners without groups (they're ignored now) - I'm going to ask design folks tomorrow on Slack and eventually update the PR/open a follow-up based on the result.
- At first I tried to add a couple of basic `ReportsLessonExerciseLearnerListPage.vue` specs related to the new feature, but got stuck on router mocks for quite a long time - I tried various ways but I always broke something else. Not sure why yet, maybe I overlooked something. After a quite long time, I gave it up for now, but I kept my test scenarios and will add them as soon as I have a look at how to simply test these views (so far it looked that `commonCoach` mixin's dependencies could cause some of those troubles so I would like to have a closer look).
- It was a bit difficult to track all dependencies to store etc. with `commonCoach` mixin when navigating in coach codes. I noticed that it's being used also for smaller components, but tried to avoid that in a new table component. I was thinking about rather splitting the mixin into helpers and smaller pieces in the future (even when it would cost us some duplicates in components/views)? Would it make sense?
- It's my first feature so would like to ask - when will be the right time to add gherkin stories? After the review?

### Reviewer guidance

1. Create a couple of groups and assign learners
2. Create a lesson containing an exercise
3. Login as some learners of groups you created and try to start/finish the exercise or stop in the middle of it
4. Visit Coach -> Reports -> Lessons -> your lesson with an exercise -> Report -> your exercise (`coach/#/<classId>/reports/lessons/<lessonId>/exercises/<exerciseId>/learners`)

- Toggle "View by groups"
- Check groups and their stats
- Check navigation between an exercise learners list and their detail views

### References

Solves first part of #5403

### Contributor Checklist

PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [x] Contributor has fully tested the PR manually
- [x] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

Commit:
Add a component to render exercise learners table

Commit:
Switch to new exercise learners table component

Commit:
Add groups toggle

Commit:
tally -> summaryTally

Commit:
Table entries can be filtered by groups

Commit:
Display tables for groups

Commit:
Navigation between grouped exercise learners list and learner detail

Commit:
exerciseLink -> exerciseLearnerLink

Commit:
Constants for last page query strings

Commit:
Show group stats

Commit:
Update statuses component doc URL

Commit:
Improve groups spacing

Commit:
Remove unused class

jonboiser commented on kolibri/plugins/coach/assets/src/views/reports/ReportsExerciseLearners.vue:
```diff
@@ -0,0 +1,108 @@
+      </tr>
+    </thead>
+    <transition-group slot="tbody" tag="tbody" name="list">
+      <tr v-for="entry in entries" ref="entry" :key="entry.id">
```
Is the `ref` used somewhere outside of this component?

MisRob replied:
Yes, in specs - I usually prefer to use `ref`s there.

jonboiser replied:
Which 'tr' does `$refs.entry` refer to? The last one?

In your test is it possible to write the test using something more high-level, like the data from vue-test-utils Wrapper methods, instead of having to access more low-level DOM info from `refs`?

MisRob replied:
Basically, it's used only to find those elements using vue-test-utils `find` - in case there is markup/css selector change, you don't need to update your `find`s, when compared to trying to find an element by a class or tag, that's all. Both ways have some advantages/disadvantages, no problem to change it if you don't use it in the codebase.

jonboiser replied:
We generally reserve the use `ref` if we need to access it in JS application code. Even if it's used in a test, it sort of looks like dead code, and I have the tendency to delete `refs` if I see they aren't used anymore in `methods`, etc.

People who want to avoid using "brittle" CSS selectors have been using this test-specific attribute `data-test`. In the spec, you are using `ref` as an attribute selector, and not really using the `ref` in the same way we use it components (e.g. low-level DOM access).

MisRob replied:
Yes, it's one of the disadvantages indeed.

I like the idea of using the attribute, it combines the best of both worlds, thanks for the tip!

jonboiser commented on kolibri/plugins/coach/assets/src/views/reports/ReportsLessonExerciseLearnerListPage.vue:
```diff
@@ -73,26 +60,78 @@
+        return this.getContentStatusTally(this.$route.params.exerciseId, recipients);
+      },
+      nonEmptyGroups() {
+        if (!this.groups || !this.groups.length) {
```
We should only show tables for groups that have been assigned the Lesson. `this.groups` will return all of the groups in the classroom.

MisRob replied:
Oh didn't notice, thanks, good catch. Will check it.

MisRob replied:
Done. I also noticed that all groups were displayed in a table in Groups column so I filtered them by lesson groups, too. Otherwise it would be misleading when switching from groups to the list of all learners.

jonboiser commented on kolibri/plugins/coach/assets/test/views/reports/ReportsExerciseLearners.spec.js:
```diff
@@ -0,0 +1,163 @@
+    const firstRow = rows.at(0);
+    const secondRow = rows.at(1);
+
+    expect(getCol(firstRow, 0).html()).toContain('learner1');
```
A snapshot test might be an efficient way to implment this particular test, since it's checking the presentation of this element.

MisRob replied:
I tried to avoid snapshot in this particular case because of "Last Activity" - not sure if mocking time would be useful only because of one test and also didn't want to test another component's responsibility but I'll check if some stub could solve it.

jonboiser replied:
If you want to be advanced, you can mock the PC clock to get a consistent "Last Activity" value.

You could also, like you say, stub 'LastActivity' so that it just doesn't appear in the snapshot, or in modified version (maybe just a span with the input date as text, for exapmle).

MisRob replied:
> If you want to be advanced, you can mock the PC clock to get a consistent "Last Activity" value

That sounds like something to be tested rather in `ElapsedTime` component.

MisRob replied:
Eventually, I added server time mock. Didn't realize that with stub, I couldn't test if an information is displayed or no.

jonboiser commented on kolibri/plugins/coach/assets/test/views/reports/ReportsExerciseLearners.spec.js:
```diff
@@ -0,0 +1,163 @@
+      entries,
+    });
+
+    const rows = wrapper.findAll({ ref: 'entry' });
```
Could you accomplish this using a CSS selector something else?

MisRob replied:
Answered above.

jonboiser commented on kolibri/plugins/coach/assets/test/views/reports/ReportsExerciseLearners.spec.js:
```diff
@@ -0,0 +1,163 @@
+    });
+
+    it("renders learner's name as a link to an exercise", () => {
+      expect(getCol(row, 0).html()).toContain('router-link');
```
You can actually stub KRouterLink so it can be treated as a very simple vue component, and the Wrapper API works on it.

Here's an example https://github.com/learningequality/kolibri/blob/develop/kolibri/plugins/coach/assets/src/views/common/notifications/__test__/NotificationCard.spec.js#L15

MisRob replied:
Not sure if I can understand this comment - why should I stub it here? I just want to check that the link is there.

jonboiser replied:
It's just an unusual test, because the actual rendered html will never have a `<router-link>` tag.

A test like
```
expect(getCol(row, 0).find({ name: 'KRouterLink' }).props().to).toEqual(...)
```

Mirrors the implementation a little more closely.

MisRob replied:
Ah, ok, get what you meant now. Yep makes sense.

indirectlylit commented:
> It was a bit difficult to track all dependencies to store etc. with commonCoach mixin when navigating in coach codes. I noticed that it's being used also for smaller components, but tried to avoid that in a new table component. I was thinking about rather splitting the mixin into helpers and smaller pieces in the future (even when it would cost us some duplicates in components/views)? Would it make sense?

That mixin was supposed to be a temporary hack and I never got around to cleaning it up. Feel free to do any cleanup you think would be appropriate.

> It's my first feature so would like to ask - when will be the right time to add gherkin stories? After the review?

You can add them any time. In theory, adding then before review means that you don't need to write reviewer guidance :)

That said, if you want to do them as a separate PR after this is merged, that's fine too - especially if you're trying to get it in before string freeze. Usually in that case I'll open a new issue specifically for the stories.

Commit:
Use data-test attr instead of ref

Commit:
Add KRouterLink stub

Commit:
Simplify specs

Commit:
Display all lesson groups

Commit:
Refactor table entries logic

to improve performance, avoid groups sources confusion
and allow easier data handling for upcoming updates
(ungrouped learners)

- save all entries to computed prop
- filter group entries from all entries instead of calculating
  everything again
- to allow things above, add group objects instead of group
  names only to table row entries
- display only lesson groups in table entries

Commit:
Display ungrouped learners

Commit:
Add specs for exercise learners lesson report page

+ discovered bugfixes

Commit:
Add a component to render resource learners table

Commit:
Use resource learners table component

Commit:
Implement groups toggling

Commit:
Refactor table entries groups

- add more granular group info to table entries to
  be able to filter it based on group id later
- include only lesson groups

Commit:
Display groups

Commit:
Add a component to display resources stats

Commit:
avgTime -> allRecipientsAvgTime

Commit:
Display group tally

Commit:
tally -> summaryTally

Commit:
Display group resources stats

MisRob commented:
@jonboiser asked me to move everything to one PR, so I'm adding it here.

Updates:
- comments resolved
- found out what was wrong with my router mocks so I could also add a couple of basic specs for exercise learners report page
- ungrouped learners are displayed in a separate section at the bottom of the page (when a whole class is a recipient)
- empty groups (that are recipients of a lesson) are displayed
- implemented same grouping feature for resources

![screencapture-localhost-8000-en-coach-2019-05-07-16_30_47](https://user-images.githubusercontent.com/13509191/57307951-b8b1ea00-70e5-11e9-9f27-5fb1c2d11096.png)
![screencapture-localhost-8000-en-coach-2019-05-07-16_31_35](https://user-images.githubusercontent.com/13509191/57307953-bc457100-70e5-11e9-9040-ac6780901fce.png)

Commit:
Add exercise/resource preview button

Commit:
Remove obsolete class

MisRob commented:
As discussed with JB, I've just also added Preview buttons to resource/exercise learners list reports to this PR:

<img width="863" alt="Screenshot 2019-05-07 at 21 25 53" src="https://user-images.githubusercontent.com/13509191/57327579-9765f380-710f-11e9-9d6e-4361ca0d8139.png">
<img width="855" alt="Screenshot 2019-05-07 at 21 26 03" src="https://user-images.githubusercontent.com/13509191/57327588-9af97a80-710f-11e9-92f4-c2f56d3e82e4.png">

This part is not exactly ideal because there are more mechanisms and in some situations, it gets a bit complicated. Also, I didn't work on a correct immersive toolbar title when navigating from reports (we agreed it should display an exercise/resource name instead of "Manage resources"). I was also thinking that maybe changing exercise/resource preview URL might be suitable after adding this feature, since it's not related to "Coach - Plan" only anymore. Anyway, we agreed that some basic functionality is OK so we can close it till Thursday.

I'll start follow-up tickets after the review and will have a look if it could be simplified a bit and also add Gherkin stories later.

When testing this PR, please try to navigate from exercise/resource list to exercise/resource preview, exercise learner detail and back for both "View by groups" checkbox states.

jonboiser commented:
For the Preview buttons, can you place them to the right of the "<- Back to Lesson Name" link. Unfortunately, the mock up was out of date with this detail.

There is a component called "BackLinkWithOptions" with that will layout the button and back link like this:

![Screen Shot 2019-05-07 at 4 13 27 PM](https://user-images.githubusercontent.com/10248067/57338750-5e178e80-70e3-11e9-89d0-ad79137d08a5.png)

jonboiser commented:
Next to the group header (h2.group-title), could you add the group icon? You can use the `<h2><KLabeledIcon><KIcon group slot="icon" /></h2>` setup.

jonboiser commented:
Don't need to fix this, but just noticed that the "Time spent" column is not being live-updated when a user finishes watching a video (not sure about exercises). Screenshot shows row for "b" after the status is live-updated to "Completed", but the timespent column is unchanged.

![Screen Shot 2019-05-07 at 4 36 42 PM](https://user-images.githubusercontent.com/10248067/57339463-85bc2600-70e6-11e9-95fe-ac8b634c9333.png)

jonboiser commented on kolibri/plugins/coach/assets/src/views/reports/ReportsLessonResourceLearnerListPage.vue:
```diff
@@ -82,12 +114,23 @@
     mixins: [commonCoach],
+    data() {
+      return {
+        viewByGroups: Boolean(this.$route.query && this.$route.query.groups),
```
I think `this.$route.query` is always going to be truthy, since it's an empty object if no query parameters are provided.

Commit:
Remove unnecessary condition

Commit:
Display group icon next to group title

Commit:
Move Preview btn next to back link

MisRob commented:
@jonboiser Just pushed all updates related to the feature, new screenshots:

![screencapture-localhost-8000-en-coach-2019-05-08-19_49_02](https://user-images.githubusercontent.com/13509191/57396399-a1006180-71ca-11e9-9e18-c429f5db8d98.png)
![screencapture-localhost-8000-en-coach-2019-05-08-19_49_36](https://user-images.githubusercontent.com/13509191/57396407-a3fb5200-71ca-11e9-8862-44eb05bb02f9.png)

jonboiser reviewed (approved):
Looks great!

MisRob commented:
Yep, I especially like the idea with group icons, it makes orientation much easier, especially when "Ungrouped learners" are present :)

MisRob commented:
Regarding "Time spent" column issue, I'll have a look at it later together with this PR's follow-ups.

jonboiser commented:
Okay, I wouldn't worry too much about the Time Spent issue. The tables are "live updated" with data coming from the Notifications API, and I'm not sure if Notifications contain the info to update the Time Spent column right now.

indirectlylit commented:
this looks awesome, great job

would one of you mind filing an issue re: the time-spent updates?

MisRob commented:
Thank you. Yes, I'm planning to have a look at it and open an issue.

MisRob commented:
#5496
